torch/csrc/jit/passes/quantization.cpp

```diff
@@ -131,43 +143,82 @@ static void addObserverFor(Value* v, Node* original_observer_node) {
 }
 
 static bool outputsNeedToBeObserved(Node* n) {
-  return n->kind() != prim::Constant;
+  return n->kind() != prim::Constant && n->kind() != prim::PythonOp;
 }
 
-void InsertObserverNodes(std::shared_ptr<Graph>& graph, Node* observer_node) {
+void InsertObserverNodes(script::Method* method,
+  std::unordered_map<std::string, Node*>& observerNodeDict) {
+  AT_ASSERT(method != nullptr);
+  auto graph = method->graph();
+  AT_ASSERT(graph != nullptr);
   // For storing all values that need to be instrumented with an observer call.
   std::vector<Value*> values_to_observe;
 
   // For traversing all blocks in the graph including subblocks.
   std::stack<Block*> blocks_to_visit;
 
+  // Observer nodes
+  Node* activation_observer = observerNodeDict.count("activation")
+    ? observerNodeDict["activation"] : nullptr;
+  Node* param_observer = observerNodeDict.count("param")
+    ? observerNodeDict["param"] : nullptr;
+
+  // Add observer for prim::Param nodes
+  auto inputVals = graph->inputs();
+  int inputlength = inputVals.size();
+  // Module params get appended after external inputs
+  int param_start_index = inputlength - method->initial_ivalues().size();
+  AT_ASSERT(param_start_index >= 0);
+  // Insert point is the beginning of graph node
+  Node* insert_node = *graph->nodes().begin();
+  for (auto idx = 0; idx < inputlength; ++idx) {
+    // This distinguish the model param from external data
+    // to insert correct observer
+    Node* observer = (idx < param_start_index) ?
+      activation_observer : param_observer;
+    if (observer == nullptr) {
+      continue;
+    }
+    auto& v = inputVals[idx];
+    if (v->type()->isSubtypeOf(TensorType::get())) {
+      addObserverFor(v, observer,
+        std::make_pair(insert_node, false));
```

Comment:
are you worried about graphs with a single node? passing this before/after is a bit ugly.

if you end up keeping it - create a tiny enum